### ISSUE TITLE
[studio] Change database upgrade mechanics

### ISSUE THREAD

**Issue by sumerjabri** [craftercms/craftercms#2355]
We need to change the upgrade mechanics to become:

1. Capable of upgrading more than just the database
2. Upgrades one version at a time in a sequence (upgrade from 1->2->3->4 instead of 1->4, 2->4, 3->4) <== this is done for the database

To achieve the above, let's do this:

- Create an upgrade manager that can upgrade any Studio 3 from any prior version to the current version
- The upgrade manager receives the existing version and the new target version and goes through a pipeline of handlers to upgrade various subsystems
- Create database upgrade pipeline that executes sequentially to upgrade from current to target, one version at a time (loop through scripts in order)

Once this is done, we will also add a new upgrade subsystem to upgrade various configuration files so we can improve the configuration files. There will be another ticket for configuration upgrades.
